driver-core/src/main/com/mongodb/client/model/expressions/DocumentExpression.java

```diff
@@ -16,10 +16,77 @@
 
 package com.mongodb.client.model.expressions;
 
+import org.bson.conversions.Bson;
+
+import java.time.Instant;
+
+import static com.mongodb.client.model.expressions.Expressions.of;
+
 /**
  * Expresses a document value. A document is an ordered set of fields, where the
  * key is a string value, mapping to a value of any other expression type.
  */
 public interface DocumentExpression extends Expression {
 
+    DocumentExpression setField(String path, Expression exp);
+
+    DocumentExpression unsetField(String path);
+
+
+    BooleanExpression getBoolean(String field);
```

Comment:
What is the behavior of `doc.getBoolean("a.b")`.  Does it 

1. return the value of the "a.b" field in this document
2. return the value of the "b" field inside the "a" field (assuming the value of the "a" field is a document)

Reply:
If 1., the parameter should be named `field`.  If 2. is there any way to use dotted path notation in this API, and if not, should there be?

Reply:
If we're not supporting dotted path notation, I worry that because their use is so prevalent in the wild, applications will make the mistake of using them and being really confused by the results they get back

Reply:
The intent was to return (1) the value of the "a.b" field, not the path.

The potential error you mention seems serious to me, but I am unsure of how to address it.

- It seems that there needs to be some way to use a literal field notation, so whatever we do we will need one `Expression get()` or many getters that take a proper field.
- We could try `getBooleanField` and `getBooleanPath`, but this seems like overkill, and perhaps would not address the possible user error.
- There is a `get(key)` on MapExpression that does not and should not take a path. MqlExpression implements both interfaces.
- If we can implement what I have been calling "schema objects" (perhaps "object bindings"?) then one can write: `doc.a.b`. I think this would disincline users from any of the getField methods, and avoid the issue, but only to the extent that object bindings are possible and used by users.

I'll keep thinking. Suggestions welcome.

Reply:
I think the more common case will be people wanting to treat the field names as paths that traverse subdocuments.  It's only in the last couple of years that the server or drivers even allowed field names with dots in them.  So that suggests that the default behavior should be to treat dotted field names as paths, and have an explicit way to treat them as field names.

At the same time, it's an attractive option to use this opportunity to "fix" the behavior, and provide explicit syntax for dotted paths.  So something like:

```
   Expression getField(String fieldName);    // treat "a.b" as a field name
   Expression getField(List<String> path);   // treat each string in the list as a field name
```

Reply:
> At the same time, it's an attractive option to use this opportunity to "fix" the behavior, and provide explicit syntax for dotted paths. So something like:
> getField(List<String> path)

If we use any type but `String` to represent field names / paths, I think we should introduce a special type instead of using `List<String>`.

I would also prefer to just introduce a `Path` type with two ways of creating its instances, something like `Path.name` and `Path.path` (in the search API we have `SearchPath.fieldPath` and `SearchPath.wildcardPath`, and we never accept plain `String` as a path in the rest of the API). We may even introduce two subtypes if there are situations in MQL when a path is not allowed and only a field name may be used (I don't know if there are such situations). I discussed this with Maxim, and he does not like the idea due to its verbosity, others may think the same. But I think that introducing a type that covers both field names (just a degenerate case of a field path) and field paths is the right way to approach the task: a plain `String` can technically be used to represent a full name of a human, text of a book, a path in a file system, a field path in MongoDB, but it does not mean that using a plain String for any of that is necessarily a good idea. Having a special type allows one to both express the meaning of a value and validate it at creation.

Reply:
Discussed, all prefer the `getInteger(String fieldName)`. It would be beneficial to have a way to query by paths, but that can be considered later.

Reply:
Added tests. "fieldName" is a fairly long parameter hint in IntelliJ, but let's see how that goes.